src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs

```diff
@@ -71,6 +80,60 @@ public void SettingUpdatesProperties()
             });
         }
 
+        [Fact]
+        public void SettingUpdatesProperties()
+        {
+            T item = GetExistingItem();
+            SettingUpdatesPropertiesCore(item);
+        }
+
+        [Theory]
+        [PlatformSpecific(~TestPlatforms.Browser)] // Browser is excluded as it doesn't support symlinks
+        [InlineData(false, false)]
+        [InlineData(false, true)]
+        [InlineData(true, false)]
+        [InlineData(true, true)]
+        public void SettingUpdatesPropertiesOnSymlink(bool targetIsRelative, bool targetExists)
+        {
+            // This test is in this class since it needs all of the time functions.
+            // This test makes sure that the times are set on the symlink itself.
+            // It is needed as on OSX for example, the default for most APIs is
+            // to follow the symlink to completion and set the time on that entry
+            // instead (eg. the setattrlist will do this without the flag set).
+            // It is also the same case on unix, with the utimensat function.
+            // It is a theory since there are variants which have a relative target
+            // or absolute target, and whether the target exists or not.
+
+            T target = targetExists ? GetExistingItem() : GetMissingItem();
+
+            // When the target exists, we want to verify that its times don't change.
+
+            T item = CreateSymlinkToItem(target, targetIsRelative);
+            if (!targetExists) SettingUpdatesPropertiesCore(item);
+            else
+            {
+                // Ensure that we have the latest times (access time could be changed by creating the symlink)
+                FileSystemInfo fsi = target as FileSystemInfo;
```

Comment:
This will be `null` when `T` is `string`. To avoid that you can add an abstract method `GetFileSystemInfo(T item)` and define how to get a FileSystemInfo for each derived class.

Reply:
Hi, the point of this code is to call `FileSystemInfo.Refresh` to invalidate the times cache before getting the times. If it uses `File` or `Directory`, it will not be cached (right?). Hence if creating a link can change these times, the cache needs to be refreshed. Also, it definitely needs to be refreshed after running the test on the link since that's the point of the extra check.

Reply:
@tmds says it will work without the first one, so I'm only keeping the 2nd one (which definitely is necessary, else it'll return the cached times). https://github.com/dotnet/runtime/pull/52639#discussion_r739761174